src/node_api.cc

```diff
@@ -17,17 +16,23 @@
 #include <vector>
 #include "node_api.h"
 
-namespace v8impl {
-
-//=== Conversion between V8 Isolate and napi_env ==========================
+napi_status napi_set_last_error(napi_env env, napi_status error_code,
+                                uint32_t engine_error_code = 0,
+                                void* engine_reserved = nullptr);
+void napi_clear_last_error(napi_env env);
 
-napi_env JsEnvFromV8Isolate(v8::Isolate* isolate) {
-  return reinterpret_cast<napi_env>(isolate);
-}
+class napi_env__ {
+ public:
+  explicit napi_env__(v8::Isolate* _isolate): isolate(_isolate), last_error() {}
+  ~napi_env__() {
+    last_exception.Reset();
+  }
+  v8::Isolate* isolate;
+  v8::Persistent<v8::Value> last_exception;
+  napi_extended_error_info last_error;
+};
```

Comment:
I'm thinkintg the base napi_env__ class should be independant of the implementation.  It would then be extended by implementations to add specifics like the isolate.  In that way the contract would be clear and we later want to stash some vm independent info in the base env we can do that and it would flow automatically into the implementations.